File: Problem_232_Implement_Queue_using_Stacks/solution.py
class MyQueue:
    def __init__(self):
        self.input_stack = []
        self.output_stack = []

    def _transfer(self):
        if not self.output_stack:
            while self.input_stack:
                self.output_stack.append(self.input_stack.pop())

    def push(self, x: int) -> None:
        self.input_stack.append(x)

    def pop(self) -> int:
        self._transfer()
        return self.output_stack.pop()

    def peek(self) -> int:
        self._transfer()
        return self.output_stack[-1]

    def empty(self) -> bool:
        return not self.input_stack and not self.output_stack


if __name__ == "__main__":
    queue = MyQueue()
    queue.push(1)
    queue.push(2)
    print(queue.peek())
    print(queue.pop())
    print(queue.empty())

File: Problem_232_Implement_Queue_using_Stacks/solution.cpp
#include <stack>
using namespace std;

class MyQueue {
public:
    void push(int x) {
        input_stack.push(x);
    }

    int pop() {
        transfer();
        int front = output_stack.top();
        output_stack.pop();
        return front;
    }

    int peek() {
        transfer();
        return output_stack.top();
    }

    bool empty() {
        return input_stack.empty() && output_stack.empty();
    }

private:
    void transfer() {
        if (output_stack.empty()) {
            while (!input_stack.empty()) {
                output_stack.push(input_stack.top());
                input_stack.pop();
            }
        }
    }

    stack<int> input_stack;
    stack<int> output_stack;
};
